### PR TITLE
ApplySchema: do not ReloadSchema on ExecuteFetchAsDba

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Followup to #10719 
There is yet another location where `ReloadSchema` is called: from within the VTTablet, and this is controlled as a parameter in  `ExecuteFetchAsDba`. We now strictly do not call `ReloadSchema` anywhere until all SQLs have been applied.

This specific PR applies to `direct` DDLs, not to Online DDLs.

## Related Issue(s)

#10719 

## Checklist

-   [ ] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added or are not required
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Review by dbussink — APPROVED** [review]

**Comment by shlomi-noach** [comment]
`Cluster (12)` consistent failure.

```
2022-07-19T10:36:39.1386134Z === RUN   TestSharded
2022-07-19T10:36:39.1563691Z     message_test.go:511: 
2022-07-19T10:36:39.1565147Z         	Error Trace:	message_test.go:511
2022-07-19T10:36:39.1566356Z         	            				message_test.go:368
2022-07-19T10:36:39.1567617Z         	Error:      	Expected nil, but got: Code: INVALID_ARGUMENT
2022-07-19T10:36:39.1570579Z         	            	rpc error: code = InvalidArgument desc = target: user.-80.primary: vttablet: rpc error: code = InvalidArgument desc = table sharded_message not found in schema (CallerID: unsecure_grpc_client)
2022-07-19T10:36:39.1573523Z         	            	target: user.80-.primary: vttablet: rpc error: code = InvalidArgument desc = table sharded_message not found in schema (CallerID: unsecure_grpc_client)
2022-07-19T10:36:39.1575588Z         	Test:       	TestSharded
2022-07-19T10:36:39.1576858Z --- FAIL: TestSharded (0.02s)
2022-07-19T10:36:39.1578162Z === RUN   TestUnsharded
2022-07-19T10:36:52.1909586Z --- PASS: TestUnsharded (13.03s)
2022-07-19T10:36:52.1911359Z === RUN   TestReparenting
2022-07-19T10:36:52.1913145Z I0719 10:36:52.190102   31342 message_test.go:600] Message stream ended: Code: CANCELED
2022-07-19T10:36:52.1915318Z rpc error: code = Canceled desc = grpc: the client connection is closing
2022-07-19T10:36:52.1943090Z     message_test.go:388: 
2022-07-19T10:36:52.1944030Z         	Error Trace:	message_test.go:388
2022-07-19T10:36:52.1945143Z         	Error:      	Expected nil, but got: Code: INVALID_ARGUMENT
2022-07-19T10:36:52.1948260Z         	            	rpc error: code = InvalidArgument desc = target: user.-80.primary: vttablet: rpc error: code = InvalidArgument desc = table sharded_message not found in schema (CallerID: unsecure_grpc_client)
2022-07-19T10:36:52.1951154Z         	            	target: user.80-.primary: vttablet: rpc error: code = InvalidArgument desc = table sharded_message not found in schema (CallerID: unsecure_grpc_client)
2022-07-19T10:36:52.1952601Z         	Test:       	TestReparenting
2022-07-19T10:36:52.1953736Z --- FAIL: TestReparenting (0.00s)
2022-07-19T10:36:52.1954919Z === RUN   TestConnection
2022-07-19T10:36:53.2056838Z     message_test.go:455: 
2022-07-19T10:36:53.2058851Z         	Error Trace:	message_test.go:455
2022-07-19T10:36:53.2060450Z         	Error:      	Expected nil, but got: Code: INVALID_ARGUMENT
2022-07-19T10:36:53.2064123Z         	            	rpc error: code = InvalidArgument desc = target: user.-80.primary: vttablet: rpc error: code = InvalidArgument desc = table sharded_message not found in schema (CallerID: unsecure_grpc_client)
2022-07-19T10:36:53.2068323Z         	            	target: user.80-.primary: vttablet: rpc error: code = InvalidArgument desc = table sharded_message not found in schema (CallerID: unsecure_grpc_client)
2022-07-19T10:36:53.2070517Z         	Test:       	TestConnection
2022-07-19T10:36:53.2072065Z --- FAIL: TestConnection (1.01s)
2022-07-19T10:36:53.2073130Z FAIL
2022-07-19T10:37:05.6304128Z FAIL	vitess.io/vitess/go/test/endtoend/messaging	52.892s
2022-07-19T10:37:05.6330069Z FAIL
```
seems related.

**Comment by shlomi-noach** [comment]
I'm lost with the test failures.

**Comment by shlomi-noach** [comment]
`vtgate_queries`, `vtgate_reservedconn` keep failing though I dont see how they're related. `vstream*` tests seem related but I can't make the connection.

**Comment by shlomi-noach** [comment]
OK I get it!
```
				schematools.ReloadShard(
					reloadCtx,
					exec.ts,
					exec.tmc,
					exec.logger,
					exec.keyspace,
					result.Shard,
					result.Position,
					concurrency,
					false, /* includePrimary */
```
says `false, /* includePrimary */`; we really do want to reload the schema on the primary though!

**Comment by shlomi-noach** [comment]
OK, tests still failing, and they're consistent, and they're persistent, and it has to do with `ApplySchema` and therefore also have to do with `ReloadSchema()`.

**Comment by shlomi-noach** [comment]
e.g. https://github.com/vitessio/vitess/runs/7413675616?check_suite_focus=true

File: go/vt/schemamanager/tablet_executor.go [hunk 1]
```diff
@@ -376,12 +376,16 @@ func (exec *TabletExecutor) Execute(ctx context.Context, sqls []string) *Execute
 		}
 	}
 
-	if syncOperationExecuted {
+	if true || syncOperationExecuted {
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/schemamanager/tablet_executor.go:379]: Yes. refactoring and also reusing your `defer` approach to ensure we `ReloadSchema` even on early return.

**Comment by shlomi-noach** [comment]
Thanks @dbussink for helping to find the root cause, which was reusing an iteration argument's address.

**Pull request merged**